src/stories/TextInput.tsx

```diff
@@ -0,0 +1,161 @@
+import React, { useState, useEffect } from "react";
+import InfoIcon from "@mui/icons-material/Info";
+import "./textinput.css";
+import TitleIcon from "@mui/icons-material/Title";
+import KeyboardReturnIcon from "@mui/icons-material/KeyboardReturn";
+
+interface TextInputProps {
+  placeholder: string;
+  isDisabled: boolean;
+  isPassword: boolean;
+  validationRegex: RegExp;
+}
+
+//This component starts by taking in placeholder, isDisabled, and isPassword props.
+//Wherever this component is used, each property in TextInputProps needs to be specified
+const TextInput = ({
+  placeholder,
+  isDisabled,
+  isPassword,
+  validationRegex,
+}: TextInputProps) => {
+  const [infoHover, setInfoHover] = useState(false);
+  const [infoMouseDown, setInfoMouseDown] = useState(false);
+
+  const [submitHover, setSubmitHover] = useState(false);
+  const [submitMouseDown, setSubmitMouseDown] = useState(false);
+
+  //realValue and setRealValue stores the actual value of the input
+  const [realValue, setRealValue] = useState("");
+  //displayValue is used to show a masked input
+  const [displayValue, setDisplayValue] = useState("");
+  //isValid is false if there are integers detected in the realValue string
+  const isValid = validationRegex.test(realValue);
+  //this hook checks if isPassword is true. Then, it either sets DisplayValue equal to either bullets or realvalue
+  //it fires whenever realValue, or isPassword changes
+
+  useEffect(() => {
+    setDisplayValue(isPassword ? "•".repeat(realValue.length) : realValue);
+  }, [realValue, isPassword]);
+
+  //this function is in charge of calculating and storing the "real value" of the input!
+  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+    //first, get the newValue from event target value
+    //e.target value looks like ••••••k, because masking logic is applied at the end
+    const newValue = e.target.value;
+    if (isPassword) {
+      if (newValue.includes("\n")) {
+        // Prevent user from entering new lines
+        return;
+      }
+      if (newValue.length > realValue.length) {
+        //Bullets from e.target.value removed
+        const addedChars = newValue.replace(/•/g, "");
+        //New character(s) added
+        setRealValue((prev) => prev + addedChars);
+      } else {
+        // Character(s) removed
+        setRealValue((prev) => prev.slice(0, newValue.length));
```

Comment:
Removing characters does not work as expected. In non-password mode, you can place the cursor anywhere then press backspace - this causes the cursor to jump to the end of the input field. Doing the same in password mode will always delete the last character. Similar behavior is observed when attempting to delete multiple characters. (Or insert one at arbitrary locations, for that matter.)